Add parachain related parameters to `chain-spec-builder`

CrackTheCode016 opened the pull request:
When using with `polkadot-parachain`, you usually need to specify the `relay_chain` and `para_id` fields in the chain spec.

With this PR it can be achieved by specifying newly added `--para-id` and `--relay-chain` command line args, e.g:
```
chain-spec-builder create -r _runtime.wasm  --para-id 100 --relay-chain xxx default
```

This was implemented by simple _json_ blobs merging.

Additionally unit tests covering basic functionality were added.

Also adds a fix for not overwriting the chain spec with the default config each time, swallowing not standard fields is also fixed.

Fixes: #4873

bkchr commented:
Generally we should start with this: https://github.com/paritytech/polkadot-sdk/issues/4873

Maybe having like a separate command to add this could be useful. I mean I'm not a 100% fan of it, but it will make it easier for the users.

CrackTheCode016 commented:
Yeah that makes sense.  I can either include a fix for that issue in this PR [(left a comment on how to potentially fix it here)](https://github.com/paritytech/polkadot-sdk/issues/4873#issuecomment-2194497860), or push another PR to address it

bkchr commented:
If you like, you can push it directly here.

bkchr commented:
@CrackTheCode016 tests are failing.

bkchr commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -158,6 +161,15 @@ pub struct CreateCmd {
 	/// The chain type.
 	#[arg(value_enum, short = 't', default_value = "live")]
 	chain_type: ChainType,
+	/// If you're generating a config for a parachain
+	#[arg(long, value_enum, default_value = "false")]
+	parachain: bool,
+	/// The para ID for your chain.
+	#[arg(value_enum, short = 'p', default_value = "100")]
+	para_id: u32,
+	/// The relay chain you wish to connect to.
+	#[arg(value_enum, short = 'c', default_value = "polkadot")]
+	relay_chain: String,
```
Why not put this into a completely different command?

michalkucharczyk replied:
I agree. Maybe something like `make_parachain`?

CrackTheCode016 replied:
Yeah that makes more sense, @michalkucharczyk would it be a subcommand to `CreateCmd` (as we want all of those args anyways for a parachain config) or a completely separate command, similar to `GenesisBuildAction`?

michalkucharczyk replied:
> This makes it easier to differentiate, otherwise, we will always have relay_chain and para_id present in all solo chain configs (set to null, i.e., "relay_chain": null).

Not sure if I understand this. I was thinking about scenario where `chain-spec-builder create` does not add any parachain related fields, and `create/make-parachain` simply injects them to chain-spec.

Sub-command is interesting idea, I like it. The only thing is that it is a little over-verbose on command line. We would need to always write `chain-spec-builder create solo ....` or `chain-spec-builder create para ....`, correct?

CrackTheCode016 replied:
We can use just plain json injection yeah.  I will try this approach.

> Sub-command is interesting idea, I like it. The only thing is that it is a little over-verbose on command line. We would need to always write chain-spec-builder create solo .... or chain-spec-builder create para ...., correct?

Yes, exactly, I am not sure how intuitive it is because of the reason you stated.  I will try your first approach because it seems simpler to the end user 👍 ty!

kianenigma replied:
I would say the overall strategy we have is to be `parachain-first`, as in, if a parameter is not specified, it should assume you are in a parachain context. With this, I suggest: 

```
chain-spec-builder create solo 
chain-spec-builder create para --relay-id x --para-id y
# should be an alias for para
chain-spec-builder create --relay-id x --para-id y
```

If with subcommands this is not possible, then I'd do the opposite: `chain-spec-builder create` builds a parachain, and `chain-spec-builder create solo` builds a solochain. 

I know this is more of a breaking change, but I think it is worthwhile. This tool is not wildly used now, but will be used a lot more in the omni-node drive future, so we better make it aligned with our standards while we can.

michalkucharczyk replied:
`chain-spec-builder create` vs `chain-spec-builder create solo` is inconsistent IMO.
`chain-spec-builder create --solo` or `chain-spec-builder create --no-para` would look better.

michalkucharczyk replied:
We also need defaults for `relay-id` and `para-id`.
If we don't want to hard-code defaults we should fail if those params are not given (and also `--solo` was not provided).

shawntabrizi replied:
I feel there is probably a way to do something "smarter" by looking at the cumulus pallets in the wasm runtime, to detect if it is a parachain or solochain

kianenigma replied:
A lot of `omni-*` tools can benefit from reading the runtime's metadata and doing checks and inference. That being said, I think the code to do it is not super trivial. Would make a great follow-up, but I would not block this PR for it, assuming the latest commits are something that majority approve.

michalkucharczyk commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -279,17 +291,40 @@ pub struct VerifyCmd {
 
 	let chain_type = &cmd.chain_type;
+	let relay_chain = &cmd.relay_chain;
+	let builder = match cmd.parachain {
```
Maybe:
```
let builder = GenericChainSpec::<Extensions>::builder(
			&code[..],
			if cmd.parachain {
			    Extensions { relay_chain: Some(relay_chain.to_string()), para_id: Some(cmd.para_id) },
             } else {
                 Exentsions::default()
             }
		)
		.with_name(&cmd.chain_name[..])
		.with_id(&cmd.chain_id[..])
		.with_chain_type(chain_type.clone()),
```

kianenigma commented on prdoc/pr_4889.prdoc:
```diff
@@ -0,0 +1,15 @@
+title: Add CLI options for parachain chain specifications + fix bug for swallowing custom fields
+
+doc:
+  - audience: Node Operator
```
oooh really? I think it is Node Dev

CrackTheCode016 replied:
I was conflicted about this, I copied some previous PRs on chain-spec-builder, you guys would know better than me about it 🙈

michalkucharczyk replied:
If it was me specifying *operator*, it is because node operator may sometime play with this util, during chain setup. But I may be wrong.
Node-dev also may never need to generate chain-spec.

CrackTheCode016 commented:
@kianenigma Yes, looking to do that tomorrow / early next week, had a few hiccups with being sick and some other tasks that needed prioritization, but will be on it asap.

Most likely will just do `create-parachain` with a couple of required fields for the relay chain, para id, and an optional field for including the "properties" for a ticker and stuff, to get the bare minimum finished for now.  

this way, anything currently using `create` will not break.

CrackTheCode016 commented:
For now, this is what I have done. Let me know if it is OK:

- I split up the core functions of generating a chain spec to be a bit more generic, so we can do stuff like solo/para builders without an issue.  
- I added the para fields as part of the `create` command, as I felt that a separate command would end up being unreasonable as it is quite repetitive. Right now, the behavior is: 
  - If no parachain fields (`--relay-chain` and `--para-id`) are provided, then it produces a normal chain-spec.  This means it is backward compatible with whoever is using it. 
  - If both `--relay-chain` and `--para-id` are provided, then a parachain chain spec is generated, along with some properties that are common + work with the template, which is what many people would be using it for.  
  
This tool could probably benefit from some more refactor for expansion later.  Something that would be nice to see is a field-by-field edit feature, so you can edit some parameters from the CLI, which I imagine is not very difficult to do. 
  
---

Solo chain spec generation: 

```sh
chain-spec-builder create \
-v \
-r runtime.wasm \
patch patch.json
```

Parachain spec generation: 

```sh
chain-spec-builder create \
-v \
--relay-chain polkadot \
--para-id 1000 \
-r runtime.wasm \
patch patch.json
```

kianenigma commented:
Your formatter seems to be at a different version than ours, which made the recent commits a bit hard to review. 

@alvicsam similar to the convo around the stable version used in CI tests, how should one find the exact nightly we use for FMT? I with that these values are stored somewhere that is intuitive to find :)

CrackTheCode016 commented:
@kianenigma Ah yes, apologies for this, I had been messing with some stuff in later versions.  Please do let me know the version to use for nightly and I'll reformat for ease of readability

skunert commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -158,6 +161,12 @@ pub struct CreateCmd {
 	#[arg(value_enum, short = 't', default_value = "live")]
 	chain_type: ChainType,
+	/// The para ID for your chain.
+	#[arg(long, value_enum, short = 'p')]
```
```suggestion
	#[arg(long, value_enum, short = 'p', requires = "relay_chain")]
```

If you add this, using `--relay-chain` cli arg require `--para-id` too, which is what we want. Using one of them without the other was silently swallowing.

skunert commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -158,6 +161,12 @@ pub struct CreateCmd {
+	#[arg(long, value_enum, short = 'p')]
+	pub para_id: Option<u32>,
+	/// The relay chain you wish to connect to.
+	#[arg(long, value_enum, short = 'c')]
```
```suggestion
	#[arg(long, value_enum, short = 'c', requires = "para_id")]
```

skunert commented:
I think this is nearly ready to merge.
There are two small things we should include, then lets go IMO.

1. When using the create command and a user only supplied `--para-id` or `--relay-chain` without the other, it is silently dropped. We can solve this at the clap level, added a comment.
2. The `convert-to-raw` command should also not swallow any fields, currently it does.

kianenigma commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -158,6 +161,12 @@ pub struct CreateCmd {
+	pub relay_chain: Option<String>,
 	/// The path to runtime wasm blob.
 	#[arg(long, short)]
 	runtime_wasm_path: PathBuf,
```
omni-bencher and chain-spec-builder, both being part of the omni-node story, should have consistent arguments. This seems one clear outlier. 

I suggest using `--runtime` everywhere. @michalkucharczyk do you agree?

michalkucharczyk replied:
> I suggest using --runtime everywhere

Makes sense. Should we just rename or deprecate `runtime_wasm_path`?

kianenigma replied:
I would do deprecation for now, emit a warning, and set it to an alias to `--runtime`.

kianenigma commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -279,18 +288,55 @@ pub struct VerifyCmd {
+	let mut properties = sc_chain_spec::Properties::new();
+	properties.insert("tokenSymbol".into(), "UNIT".into());
+	properties.insert("tokenDecimals".into(), 12.into());
+	properties.insert("ss58Format".into(), 42.into());
```
I am pretty sure this is not used by anyone, and is now part of `frame_system::Config`.

CrackTheCode016 replied:
I still see chain specs using this, so maybe this is a recent thing - I tried looking but could not find it, any chance you could share a link?

CrackTheCode016 replied:
I tried looking for these parameters in `frame_system::Config` but could not find them.  I guess if we do include them, it would be appropriate to offer flags for them in the CLI anyway, but I can remove them for now

kianenigma replied:
https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/trait.Config.html#associatedtype.SS58Prefix

michalkucharczyk replied:
removed

michalkucharczyk commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -279,18 +288,55 @@ pub struct VerifyCmd {
+	properties.insert("tokenDecimals".into(), 12.into());
+	properties.insert("ss58Format".into(), 42.into());
+
+	GenericChainSpec::<ParachainExtension>::builder(
```
We don't really need extension and generic type parameter here. We could simply merge two json blobs using (json merge)[https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/json_patch/fn.merge.html] in case we generate for para.

But for now it is fine.

CrackTheCode016 replied:
So essentially create two chain specs (our "custom" para one then a generic one), then merge them?

michalkucharczyk replied:
I've changed this to merging jsons.

michalkucharczyk reviewed (approved):
Looks good to me.

I think swallowing fields in `ConvertToRaw` can be done as separate PR.
Would be nice to have test-suit covering most typical scenarios (which now are only in documentation), but this is also out of scope of this PR.

CrackTheCode016 commented:
Regarding this issue:

> I think swallowing fields in ConvertToRaw can be done as separate PR.

I am actually not sure what is causing this. I thought initially, it had to do with the fact that `GenericChainSpec<()>` indicated no extensions, but it's simply a generic parameter which serde flattens anyway.  It seems that serde disregards any unknown fields in ClientSpec, even though `#[serde(deny_unknown_fields)]` is no where that seems relevant, but I could be missing something.

michalkucharczyk commented:
> Regarding this issue:
> 
> > I think swallowing fields in ConvertToRaw can be done as separate PR.
> 
> I am actually not sure what is causing this. I thought initially, it had to do with the fact that `GenericChainSpec<()>` indicated no extensions, but it's simply a generic parameter which serde flattens anyway. It seems that serde disregards any unknown fields in ClientSpec, even though `#[serde(deny_unknown_fields)]` is no where that seems relevant, but I could be missing something.
> 
> Either way, it makes sense to conclude this PR, and add another to fix this issue.

This structure is responsible for reading chain spec:
https://github.com/paritytech/polkadot-sdk/blob/7e7c33453eeb14f47c6c4d0f98cc982e485edc77/substrate/client/chain-spec/src/chain_spec.rs#L264-L297

It does not have  `#[serde(deny_unknown_fields)]` , so it silently ignores all unknown fields when reading file, and serializes only struct fields when creating new json blob. As Basti mentioned we need to keep original json, and merge it to the output json blob.

kianenigma commented on substrate/bin/utils/chain-spec-builder/bin/main.rs:
```diff
@@ -46,18 +55,19 @@ fn inner_main() -> Result<(), String> {
-			let chain_spec_json = generate_chain_spec_for_runtime(&cmd)?;
+			let is_para = cmd.relay_chain.clone().zip(cmd.para_id.clone());
+			let chain_spec_json = match is_para {
+				Some(_) => generate_chain_spec_for_para_runtime(&cmd)?,
```
A little rust nit: 

You assert that `cmd.relay_chain` and `cmd.para_id` are both `Some(_)` in this function, and then the `.expect` is in the `generate_chain_spec_for_para_runtime`. A nicer way would be to pass the unwrapped values inside the `Some(_)` to this function to make it more read-able.

michalkucharczyk replied:
I reworked this piece:
https://github.com/paritytech/polkadot-sdk/blob/ce556f19e37c358a12204d51d58a4251633d7c24/substrate/bin/utils/chain-spec-builder/src/lib.rs#L480-L489

I also got rid of newly introduced `Extensions` for parachains. IMO merging jsons is simpler, we also have the same chain spec generation flow for solo and para.

kianenigma reviewed (commented):
Would be happy to approve this, but some of previous rounds of comments are not finished yet.

@CrackTheCode016 we could also discuss handing over this task to someone else, if it would help you :)

CrackTheCode016 commented:
@kianenigma I think that would be a good idea, I have some priorities at the moment and someone could probably come along and get these done within a day.  

The changes are pretty minute - so it shouldn't be a big deal.  I think really the main one is merging the JSON so non-struct fields don't get eaten up.

michalkucharczyk commented:
I will take it from here.

kianenigma commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -220,7 +233,8 @@ pub struct UpdateCodeCmd {
 	pub input_chain_spec: PathBuf,
 	/// The path to new runtime wasm blob to be stored into chain-spec.
-	pub runtime_wasm_path: PathBuf,
+	#[arg(alias = "runtime-wasm-path")]
```
Thank you!

kianenigma commented:
Thank you @michalkucharczyk!

From my side, I'd mainly like to see the discussed guideline of "it should feel like omni-bencher" fulfilled. AFAIK this is already done, so will be ready for a final review here :)

michalkucharczyk commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -279,18 +288,55 @@ pub struct VerifyCmd {
-	let builder = GenericChainSpec::<()>::builder(&code[..], Default::default())
+	let mut properties = sc_chain_spec::Properties::new();
+	properties.insert("tokenSymbol".into(), "UNIT".into());
+	properties.insert("tokenDecimals".into(), 12.into());
```
Now it is consistent.
Quesion is: shall we keep it? I think we don't need it.

michalkucharczyk commented:
I'd say this is ready for next round of reviews. 

I also fixed swallowing unknown fields, applied all the comments from previous reviews (hope I did not miss anything), and also added some testsuite for basic scenarios.

skunert commented on substrate/bin/utils/chain-spec-builder/tests/test.rs:
```diff
@@ -0,0 +1,186 @@
+use staging_chain_spec_builder::ChainSpecBuilder;
+
+const RUNTIME_PATH: &str =
+	"../../../../target/release/wbuild/substrate-test-runtime/substrate_test_runtime.wasm";
```
Can we assume this is always present? When I just execute `cargo test` for this crate, the tests fail because this file is missing.

skunert replied:
My bad, I was not passing `--release`. Would have been nice to support debug tests too, but if its more as a hassle I think its okay.

michalkucharczyk replied:
we could pass wasm blob to the run function. I'll improve it.

skunert commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -146,7 +153,7 @@ pub enum ChainSpecBuilderCmd {
 }
 
-/// Create a new chain spec by interacting with the provided runtime wasm blob.
+/// Create a new solo chain spec by interacting with the provided runtime wasm blob.
```
```suggestion
/// Create a new chain spec by interacting with the provided runtime wasm blob.
```
I mean we now support parachains :D

skunert commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -279,18 +294,130 @@ pub struct VerifyCmd {
+						)
+					})
+					.collect();
+				println!("{}", serde_json::json!({"presets":presets}).to_string());
```
Nit: I am not sure if this should be a json. Intuitively, I would expect that a cli tool just returns the name in a plain list. But its more a gut feeling, no hard opinion.

skunert commented on substrate/bin/utils/chain-spec-builder/src/lib.rs:
```diff
@@ -279,18 +294,130 @@ pub struct VerifyCmd {
+						.map_err(|e| format!("Conversion to json failed: {e}"))?;
+
+				// We want to extract only raw genesis ("genesis::raw" key), and apply it as a patch
+				// for the original json file. However, the file also contains origianl plain
```
```suggestion
				// for the original json file. However, the file also contains original plain
```

skunert reviewed (commented):
Looks really good, only some nits.
